File: Kotlin/0002.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun addTwoNumbers(l1: ListNode?, l2: ListNode?): ListNode? {
        // Create a dummy node to simplify operations
        val dummyHead = ListNode(0)
        var current = dummyHead
        var carry = 0

        var p = l1
        var q = l2

        // Traverse both linked lists
        while (p != null || q != null || carry != 0) {
            // Get the current values
            val x = p?.`val` ?: 0
            val y = q?.`val` ?: 0

            // Calculate the sum of current digits and carry
            val sum = x + y + carry
            carry = sum / 10
            val newVal = sum % 10

            // Create a new node with the computed value
            current.next = ListNode(newVal)
            current = current.next!!

            // Move to the next nodes in the lists
            p = p?.next
            q = q?.next
        }

        // Return the next of dummy node, which is the head of the result list
        return dummyHead.next
    }
}

File: PHP/0002.php
/**
 * Definition for a singly-linked list.
 * class ListNode {
 *     public $val = 0;
 *     public $next = null;
 *     function __construct($val = 0, $next = null) {
 *         $this->val = $val;
 *         $this->next = $next;
 *     }
 * }
 */
class Solution {

    /**
     * @param ListNode $l1
     * @param ListNode $l2
     * @return ListNode
     */
    function addTwoNumbers($l1, $l2) {
        // Create a dummy node to simplify operations
        $dummyHead = new ListNode(0);
        $current = $dummyHead;
        $carry = 0;

        // Traverse both linked lists
        while ($l1 !== null || $l2 !== null || $carry !== 0) {
            // Get the current values
            $val1 = ($l1 !== null) ? $l1->val : 0;
            $val2 = ($l2 !== null) ? $l2->val : 0;

            // Calculate the sum of current digits and carry
            $sum = $val1 + $val2 + $carry;
            $carry = intdiv($sum, 10); // Update carry
            $newVal = $sum % 10; // Get the last digit of sum

            // Create a new node with the computed value
            $current->next = new ListNode($newVal);
            $current = $current->next;

            // Move to the next nodes in the lists
            if ($l1 !== null) $l1 = $l1->next;
            if ($l2 !== null) $l2 = $l2->next;
        }

        // Return the next of dummy node, which is the head of the result list
        return $dummyHead->next;
    }
}